model/fn-execution/src/main/proto/beam_fn_api.proto

```diff
@@ -257,6 +262,122 @@ message ProcessBundleProgressRequest {
 
+message MonitoringInfo {
+  // The name defining the metric or monitored state.
+  string urn = 1;
+
+  // This is specified as a URN that implies:
+  // A message class: (Distribution, Counter, Extrema, MonitoringDataTable).
+  // Sub types like field formats - int64, double, string.
+  // Aggregation methods - SUM, LATEST, TOP-N, BOTTOM-N, DISTRIBUTION
+  // valid values are:
+  // beam:metrics:[SumInt64|LatestInt64|Top-NInt64|Bottom-NInt64|
+  //     SumDouble|LatestDouble|Top-NDouble|Bottom-NDouble|DistributionInt64|
+  //     DistributionDouble|MonitoringDataTable]
+  string type = 2;
+
+  // The Metric or monitored state.
+  oneof monitoring_status {
+    MonitoringTableData monitored_table_data = 3;
+    Metric metric = 4;
+  }
+
+  // A set of key+value labels which define the scope of the metric.
+  // Either a well defined entity id for the keys:
+  // “transform”, “pcollection”, “windowing_strategy”,
+  // “coder”, “environment” or any arbitrary label
+  // set by a custom metric or user metric.
+  // A monitoring system is expected to be able to aggregate the metric together
+  // for all updates having the same URN and labels.
+  // Some systems such as Stackdriver will be able to aggregate the metric
+  // using a subset of the provided labels
+  map<string, string> labels = 5;
+}
+
+message Metric {
+  // (Required) The data for this metric.
+  oneof data {
+    CounterData counter_data = 1;
+    DistributionData distribution_data = 2;
+    Extrema extrema_data = 3;
+  }
+}
+
+// Data associated with a Counter or Gauge metric.
+// This is designed to be compatible with metric collection
+// systems such as DropWizard.
+message CounterData {
+   oneof value {
+     int64 int64_value = 1;
+     string string_value = 2;
+     double double_value = 3;
+   }
+}
+
+// Extrema messages are used for calculating
+// Top-N/Bottom-N metrics.
+message Extrema {
+  // Only one of the two should be specified.
+  // Note: oneof is not allowed on repeated fields.
+  repeated int64 int_values = 1;
+  repeated double double_values = 2;
+}
+
+// Data associated with a distribution metric.
+// This is based off of the current DistributionData metric
+// This is not a stackdriver or dropwizard compatible
+// style of distribution metric.
+message DistributionData {
+  oneof distribution {
+    IntDistributionData int_double_distribution = 1;
+    DoubleDistributionData double_distribution_data =2;
+  }
+}
+
+message IntDistributionData {
+  int64 count = 1;
+  int64 sum = 2;
+  int64 min = 3;
+  int64 max = 4;
+}
+
+message DoubleDistributionData {
+  int64 count = 1;
+  double sum = 2;
+  double min = 3;
+  double max = 4;
+}
+
+// General MonitoredState information which contains
+// structured information which does not fit into a typical
+// metric format. For example, a table of important files
+// and metadata which an I/O source is reading.
+// Note: Since MonitoredState is designed to be
+// customizable, and allow engines to aggregate these
```

Comment:
engines don't customize them as below you mention that the aggregation is always just latest from the runners perspective.

Reply:
Latest across all shards/bundles? Or Union?

Reply:
Perhaps I did not describe this well. Updating... What I mean to say is that a custom URN can do a custom aggregation, if an engine choose to support it in its aggregation system.

Consider the I/O source reading files, emitting the table of file statuses:
The workers, will do the same thing always, just emit their 'current state'
I.e. every time I emit my oldest three file which have been waiting for data for over X hours.
Then if the engine supports the particular URN and has custom handling it can just maintain the Top-N oldest files.

Otherwise, it can just union all the information together and create a large MonitoringTableData (agnostic of what is inside it).